data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/ScanObjectWorker.java

```diff
@@ -36,118 +37,126 @@ public class ScanObjectWorker implements Runnable{
+            sourceCoordinator.giveUpPartitions();
         }
     }
-    /**
-     * Method will identify already processed key.
-     * @return boolean
-     */
-    private boolean isKeyNotProcessedByS3Scan(final S3ObjectDetails s3ObjectDetails) {
-        return stateSaveMap.get(s3ObjectDetails.getBucket()+s3ObjectDetails.getKey()) == null;
+
+    private List<PartitionIdentifier> provideFilteredBucketKeyPartitions() {
+        final List<PartitionIdentifier> objectsToProcess = new ArrayList<>();
+
+        for (final ScanOptions scanOptions : scanOptionsBuilderList) {
+            final List<String> excludeItems = new ArrayList<>();
+            final S3ScanKeyPathOption s3ScanKeyPathOption = scanOptions.getS3ScanKeyPathOption();
+            final ListObjectsV2Request.Builder listObjectsV2Request = ListObjectsV2Request.builder()
+                    .bucket(scanOptions.getBucket());
+            bucketOwnerProvider.getBucketOwner(scanOptions.getBucket())
+                    .ifPresent(listObjectsV2Request::expectedBucketOwner);
+
+            if (Objects.nonNull(s3ScanKeyPathOption) && Objects.nonNull(s3ScanKeyPathOption.getS3ScanExcludeSuffixOptions()))
+                excludeItems.addAll(s3ScanKeyPathOption.getS3ScanExcludeSuffixOptions());
+
+            if (Objects.nonNull(s3ScanKeyPathOption) && Objects.nonNull(s3ScanKeyPathOption.getS3scanIncludeOptions()))
+                s3ScanKeyPathOption.getS3scanIncludeOptions().forEach(includePath -> {
+                    listObjectsV2Request.prefix(includePath);
+                    objectsToProcess.addAll(listS3Objects(excludeItems, listObjectsV2Request, scanOptions.getBucket(),
+                            scanOptions.getUseStartDateTime(), scanOptions.getUseEndDateTime()));
+                });
+            else
+                objectsToProcess.addAll(listS3Objects(excludeItems, listObjectsV2Request, scanOptions.getBucket(),
+                        scanOptions.getUseStartDateTime(), scanOptions.getUseEndDateTime()));
+        }
+
+        return objectsToProcess;
     }
 
-    /**
-     * store the processed bucket key in the map.
-     */
-    private void updateKeyProcessedByS3Scan(final S3ObjectDetails s3ObjectDetails) {
-        stateSaveMap.put((s3ObjectDetails.getBucket() + s3ObjectDetails.getKey()),s3ObjectDetails);
+    private List<PartitionIdentifier> listS3Objects(final List<String> excludeKeyPaths,
+                                                    final ListObjectsV2Request.Builder listObjectsV2Request,
+                                                    final String bucket,
+                                                    final LocalDateTime startDateTime,
+                                                    final LocalDateTime endDateTime) {
+          return s3Client.listObjectsV2(listObjectsV2Request.fetchOwner(true).build()).contents().stream()
```

Comment:
Does this have built-in paging? How do we get all the objects?

Reply:
Looking at it this will get up to 1,000 objects

Reply:
Thanks for taking note of that. We can make a GitHub issue for this and resolve as a follow-on.

Reply:
I included the pagination in the next revision. If more work is needed I can create a separate issue for it.